@()(implicit request: play.api.mvc.Request[Any])

<!DOCTYPE html>


<html>
    <head>
        @head("GMQL")
        @*My less*@
    <link rel="stylesheet" href="@routes.Assets.at("stylesheets/login.css")">
    <link rel="stylesheet" href="@routes.Assets.at("stylesheets/main.css")">
    <link rel="stylesheet" href="@routes.Assets.at("stylesheets/tree.css")">

    </head>
    <body data-spy="scroll" data-target=".bs-docs-sidebar">
        <div>@views.html.gmql.gmql_navbar()</div>
        <div>@views.html.gmql.register_modal()</div>



            <!-- Main Content -->
        <div class="row MainContent">
                <!--Side Nav Bar -->
            <nav class="col-xs-3 bs-docs-sidebar">
                <ul id="sidebar" class="nav nav-stacked fixed">
                    <li>
                        <a href="#GroupA">Group A</a>
                        <ul class="nav nav-stacked">
                            <li><a href="#GroupASub1">Sub-Group 1</a></li>
                            <li><a href="#GroupASub2">Sub-Group 2</a></li>
                        </ul>
                    </li>
                    <li>
                        <a href="#GroupB">Group B</a>
                        <ul class="nav nav-stacked">
                            <li><a href="#GroupBSub1">Sub-Group 1</a></li>
                            <li><a href="#GroupBSub2">Sub-Group 2</a></li>
                        </ul>
                    </li>
                    <li>
                        <a href="#GroupC">Group C</a>
                        <ul class="nav nav-stacked">
                            <li><a href="#GroupCSub1">Sub-Group 1</a></li>
                            <li><a href="#GroupCSub2">Sub-Group 2</a></li>
                        </ul>
                    </li>
                </ul>
            </nav>

                <!--Main Content -->
            <div class="col-xs-9">
                <section id="GroupA" class="group">
                    <h3 class="sectionHead"><span class="titlemark">1   </span> <a
                    id="x1-20001"></a>Introduction and Motivations</h3>
                        <!--l. 46--><p class="noindent" >
                    A new technology for reading the DNA, called Next Generation Sequencing (NGS),
                    is changing biological research and will change medical practice, thanks to the
                    low-cost availability of millions of whole genome sequences of a variety of species,
                    and most important of humans. Huge repositories of sequence information are being
                    collected by large consortia of research laboratories by using NGS; among them,
                    ENCODE <span class="cite">[<a
                href="#X[ENCODE]"
onmouseover="tex4ht.body = tex4ht.XjVVENCODEjVV;
tex4ht.caption = '[ENCODE]';
return overlib(tex4ht.body, FULLHTML);"
onmouseout="nd();"
                >12</a>]</span>, TCGA <span class="cite">[<a
                href="#X[TCGA]"
onmouseover="tex4ht.body = tex4ht.XjVVTCGAjVV;
tex4ht.caption = '[TCGA]';
return overlib(tex4ht.body, FULLHTML);"
onmouseout="nd();"
                >35</a>]</span>, the <span
                class="cmti-10x-x-109">1000 Genomes Project </span><span class="cite">[<a
                href="#X1000"
onmouseover="tex4ht.body = tex4ht.X1000;
tex4ht.caption = '1000';
return overlib(tex4ht.body, FULLHTML);"
onmouseout="nd();"
                >1</a>]</span> and the <span
                class="cmti-10x-x-109">100,000 Genomes</span>
                    <span
                    class="cmti-10x-x-109">
                        Project</span><span class="footnote-mark"><!--\def\jstoday{2016-09-20 18:01:00} \input onmouseover.4ht   <js>-->  <a
                href="#fn1x0" id="fn1x0-bk"
onmouseover="tex4ht.body = tex4ht.fn1x0;
tex4ht.caption = '1';
return overlib(tex4ht.body, FULLHTML);"
onmouseout="nd();"

                ><sup class="textsuperscript">1</sup></a></span><a
                id="x1-2001f1"></a>.
                    These sequences can be assembled with specific experimental data produced at
                    the various research or clinical centers, opening new opportunities for biological
                    discovery and for personalized medicine.
                </p><!--l. 55--><p class="indent" >
                    Several organizations are considering genomics at a global level. <span
                class="cmti-10x-x-109">Global Alliance for genomics</span>
                    <span
                    class="cmti-10x-x-109">and Health</span><span class="footnote-mark"><a
                href="#fn2x0" id="fn2x0-bk"
onmouseover="tex4ht.body = tex4ht.fn2x0;
tex4ht.caption = '2';
return overlib(tex4ht.body, FULLHTML);"
onmouseout="nd();"

                ><sup class="textsuperscript">2</sup></a></span><a
                id="x1-2002f2"></a>
                    is a large consortium of over 200 research institutions with the goal of
                    supporting voluntary and secure sharing of genomic and clinical data;
                    their work on data interoperability is producing a data conversion
                    technology<span class="footnote-mark"><a
                href="#fn3x0" id="fn3x0-bk"
onmouseover="tex4ht.body = tex4ht.fn3x0;
tex4ht.caption = '3';
return overlib(tex4ht.body, FULLHTML);"
onmouseout="nd();"

                ><sup class="textsuperscript">3</sup></a></span><a
                id="x1-2003f3"></a>.
                    <span
                    class="cmti-10x-x-109">Google </span>recently provided an API to store, process, explore, and share
                    DNA sequence reads, alignments, and variant calls, using Google&#8217;s cloud
                    infrastructure<span class="footnote-mark"><a
                href="#fn4x0" id="fn4x0-bk"
onmouseover="tex4ht.body = tex4ht.fn4x0;
tex4ht.caption = '4';
return overlib(tex4ht.body, FULLHTML);"
onmouseout="nd();"

                ><sup class="textsuperscript">4</sup></a></span><a
                id="x1-2004f4"></a>.
                    Parallel frameworks are used to support genomic computing, including <span
                class="cmti-10x-x-109">Vertica</span>
                    (used by Broad Institute and NY Genome Center) and <span
                class="cmti-10x-x-109">SciDB </span>(used by NCBI for
                    storing the data of the 1000 Genome project <span class="cite">[<a
                href="#X1000"
onmouseover="tex4ht.body = tex4ht.X1000;
tex4ht.caption = '1000';
return overlib(tex4ht.body, FULLHTML);"
onmouseout="nd();"
                >1</a>]</span>).
                </p><!--l. 58--><p class="indent" >
                    So far, the bio-informatics research community has been mostly challenged by
                    <span
                    class="cmti-10x-x-109">primary analysis </span>(production of sequences in the form of shorts DNA
                    segments, or &#8221;reads&#8221;) and <span
                class="cmti-10x-x-109">secondary analysis </span>(alignment of reads to a
                    reference genome and search for specific features, such as variants/mutations
                    and peaks of expression); but the most important emerging problem is
                    the so-called <span
                class="cmti-10x-x-109">tertiary analysis</span>, concerned with multi-sample processing,
                    annotation and filtering of variants, and genome browser-driven exploratory
                    analysis <span class="cite">[<a
                href="#X[Paradigm4]"
onmouseover="tex4ht.body = tex4ht.XjVVParadigm4jVV;
tex4ht.caption = '[Paradigm4]';
return overlib(tex4ht.body, FULLHTML);"
onmouseout="nd();"
                >25</a>]</span>. While secondary analysis targets <span
                class="cmti-10x-x-109">raw data </span>in output from
                    NGS processors by using specialized methods, tertiary analysis targets
                    <span
                    class="cmti-10x-x-109">processed data </span>
                    in output from secondary analysis and is responsible of <span
                class="cmti-10x-x-109">sense</span>
                    <span
                    class="cmti-10x-x-109">making</span>, e.g., discovering how heterogeneous regions interact with each
                    other.
                </p><!--l. 68--><p class="indent" >   The <span
                class="cmbx-10x-x-109">GenData 2020 </span>research project was conceived to address
                    this challenge, by enabling queries and analysis of processed genomic
                    data<span class="footnote-mark"><a
                href="#fn5x0" id="fn5x0-bk"
onmouseover="tex4ht.body = tex4ht.fn5x0;
tex4ht.caption = '5';
return overlib(tex4ht.body, FULLHTML);"
onmouseout="nd();"

                ><sup class="textsuperscript">5</sup></a></span><a
                id="x1-2005f5"></a>.


                    The project&#8217;s main results so far are a <span
                class="cmti-10x-x-109">Genomic Data Mode</span>l (GDM), which
                    encodes processed data in terms of their regions and metadata, and a
                    <span
                    class="cmti-10x-x-109">GenoMetric Query Language </span>(GMQL) for extracting regions of interest
                    from experiments and for computing their properties, with high-level
                    operations for manipulating regions and for measuring their distances
                    <span class="cite">[<a
                    href="#X[GMQL]"
onmouseover="tex4ht.body = tex4ht.XjVVGMQLjVV;
tex4ht.caption = '[GMQL]';
return overlib(tex4ht.body, FULLHTML);"
onmouseout="nd();"
                    >20</a>]</span>.
                </p><!--l. 75--><p class="indent" >
                    The most relevant and, to the best of our knowledge, original aspects of
                    <span
                    class="cmbx-10x-x-109">GenData 2020 </span>
                    is the targeting of the project towards heterogeneous processed
                    data rather than raw data. World-wide genomic repositories already contain huge
                    amounts of processed data, and actually their value stems from the certification
                    of high-quality processing. While processed data are much smaller than raw data,
                    they can be considered as &#8220;big data&#8221;, because each processed file can contain
                    thousands or even millions of genomic regions. In the <span
                class="cmbx-10x-x-109">GenData 2020 </span>repository
                    structure, we show that thousands of files can be extracted from the repositories
                    and organized within one dataset, that can be simply referenced by name in
                    GMQL.
                </p><!--l. 84--><p class="indent" >   Another unique aspect of <span
                class="cmbx-10x-x-109">GenData 2020 </span>is the inclusion of metadata in the
                    GDM model and of metadata management in the GMQL query language.
                    Each dataset includes in its metadata all known information about each
                    sample (from the sample&#8217;s preparation up to the patient&#8217;s phenotype.)
                    GMQL progressively computes both the regions and the metadata of
                    resulting samples. Thus, result samples of every query carry their meta-data,
                    linking them to their contributing input samples, as an indication of <span
                class="cmbx-10x-x-109">data</span>
                    <span
                    class="cmbx-10x-x-109">provenance</span>
                    ; this is very powerful, e.g. for building genotype-phenotype
                    associations.
                </p><!--l. 93--><p class="indent" >   The <span
                class="cmbx-10x-x-109">GenData 2020 </span>implementation uses cloud computing. Specifically,
                    Version 1 of the system, developed between 2013 and 2014, translates
                    GMQL to Pig <span class="cite">[<a
                href="#X[Pig-APACHE]"
onmouseover="tex4ht.body = tex4ht.XjVVPigJjVAPACHEjVV;
tex4ht.caption = '[Pig-APACHE]';
return overlib(tex4ht.body, FULLHTML);"
onmouseout="nd();"
                >6</a>]</span> in the context of Hadoop 1 HDFS, while Version 2 of the
                    system, developed in 2015, uses the Spark <span class="cite">[<a
                href="#X[Spark-APACHE]"
onmouseover="tex4ht.body = tex4ht.XjVVSparkJjVAPACHEjVV;
tex4ht.caption = '[Spark-APACHE]';
return overlib(tex4ht.body, FULLHTML);"
onmouseout="nd();"
                >7</a>]</span> and Flink <span class="cite">[<a
                href="#X[Flink-APACHE]"
onmouseover="tex4ht.body = tex4ht.XjVVFlinkJjVAPACHEjVV;
tex4ht.caption = '[Flink-APACHE]';
return overlib(tex4ht.body, FULLHTML);"
onmouseout="nd();"
                >4</a>]</span> frameworks as
                    supported within Hadoop YARN <span class="cite">[<a
                href="#X[Hadoop 2]"
onmouseover="tex4ht.body = tex4ht.XjVVHadoop2jVV;
tex4ht.caption = '[Hadoop 2]';
return overlib(tex4ht.body, FULLHTML);"
onmouseout="nd();"
                >16</a>]</span>; the need of cloud architectures for
                    genomics is advocated by <span class="cite">[<a
                href="#XNEW"
onmouseover="tex4ht.body = tex4ht.XNEW;
tex4ht.caption = 'NEW';
return overlib(tex4ht.body, FULLHTML);"
onmouseout="nd();"
                >30</a>]</span>. The expressive power and flexibility of
                    GenData 2020&#8217;s data model (GDM) and query language (GMQL) are
                    demonstrated in <span class="cite">[<a
                href="#X[GMQL]"
onmouseover="tex4ht.body = tex4ht.XjVVGMQLjVV;
tex4ht.caption = '[GMQL]';
return overlib(tex4ht.body, FULLHTML);"
onmouseout="nd();"
                >20</a>]</span>, where we show four very different genomic use
                    cases<span class="footnote-mark"><a
                href="#fn6x0" id="fn6x0-bk"
onmouseover="tex4ht.body = tex4ht.fn6x0;
tex4ht.caption = '6';
return overlib(tex4ht.body, FULLHTML);"
onmouseout="nd();"

                ><sup class="textsuperscript">6</sup></a></span><a
                id="x1-2006f6"></a>. GenData
                    2020 focuses on tertiary data analysis; a similar approach is advocated by <span
                class="cmti-10x-x-109">Paradigm4</span>, a
                    company <span class="footnote-mark"><a
                href="#fn7x0" id="fn7x0-bk"
onmouseover="tex4ht.body = tex4ht.fn7x0;
tex4ht.caption = '7';
return overlib(tex4ht.body, FULLHTML);"
onmouseout="nd();"

                ><sup class="textsuperscript">7</sup></a></span><a
                id="x1-2007f7"></a>
                    whose products include genomic adds-on to SciDB <span class="cite">[<a
                    href="#X[Paradigm4]"
onmouseover="tex4ht.body = tex4ht.XjVVParadigm4jVV;
tex4ht.caption = '[Paradigm4]';
return overlib(tex4ht.body, FULLHTML);"
onmouseout="nd();"
                    >25</a>]</span>. They also concentrate on
                    the <span
                    class="cmti-10x-x-109">tertiary analysis</span>
                    , but they advocate the use of specialized databases rather
                    than cloud computing.
                    <!-- -->





























                <div id="GroupASub1" class="subgroup">
                    <h4 class="subsectionHead"><span class="titlemark">2.1   </span> <a
                    id="x1-40002.1"></a>Motivation</h4>
                        <!--l. 115--><p class="noindent" >Processed data have a variety of file formats, and typically lack an attribute-based
                    organization. GDM provides a schema to regions, thus it makes data
                    self-describing, as advocated by Jim Gray <span class="cite">[<a
                href="#X[Gray]"
onmouseover="tex4ht.body=tex4ht.XjVVGrayjVV; tex4ht.caption='[Gray]'; return overlib(tex4ht.body,FULLHTML);"
onmouseout="nd();"
                >15</a>]</span>; however, we don&#8217;t include data
                    into a database, so as to preserve the possibility for biologists to work with their
                    usual file-based tools.
                </p><!--l. 120--><p class="indent" >   The schema has a fixed part that guarantees the comparability of regions produced
                    by different kinds of processing, and then a variable part describing the features
                    produced by the various kinds of processing. Although DNA regions are strings of
                    nucleotides<span class="footnote-mark"><a
                href="#fn8x0" id="fn8x0-bk"
onmouseover="tex4ht.body=tex4ht.fn8x0; tex4ht.caption='8'; return overlib(tex4ht.body,FULLHTML);"
onmouseout="nd();"

                ><sup class="textsuperscript">8</sup></a></span><a
                id="x1-4001f8"></a>,
                    we instead associate them with a list of one or more <span
                class="cmti-10x-x-109">features</span>, where each feature
                    is produced by secondary data analysis.
                </p><!--l. 128--><p class="indent" >   Due to the lack of agreed standards for metadata, we model them as free
                    attribute-value pairs; we expect metadata to include at least the experiment
                    type, the sequencing and analysis method used for data production, the cell line,
                    tissue, experimental condition (e.g., antibody target) and organism sequenced; in
                    case of clinical studies, individual&#8217;s descriptions including phenotypes. Attributes
                    may have multiple values (e.g., the <span
                class="cmtt-10x-x-109">Disease </span>attribute can have both
                    values <span
                class="cmtt-10x-x-109">&#8217;Cancer&#8217; </span>and <span
                class="cmtt-10x-x-109">&#8217;Diabetes&#8217;</span>). Hundreds of datasets and thousands of
                    samples<span class="footnote-mark"><a
                href="#fn9x0" id="fn9x0-bk"
onmouseover="tex4ht.body=tex4ht.fn9x0; tex4ht.caption='9'; return overlib(tex4ht.body,FULLHTML);"
onmouseout="nd();"

                ><sup class="textsuperscript">9</sup></a></span><a
                id="x1-4002f9"></a>
                    can be queried thanks to the GDM model.
                </p>
                </div>









                    <div id="GroupASub2" class="subgroup">
                        <h4>Group A Sub 2</h4>
                        <br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br>
                        qwe
                        <br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br>

                    </div>
                </section>
                <section id="GroupB" class="group">
                    <h3>Group B</h3>
                    <div id="GroupBSub1" class="subgroup">
                        <h4>Group B Sub 1</h4>
                    </div>
                    <div id="GroupBSub2" class="subgroup">
                        <h4>Group B Sub 2</h4>
                    </div>
                </section>
                <section id="GroupC" class="group">
                    <h3>Group C</h3>
                    <div id="GroupCSub1" class="subgroup">
                        <h4>Group C Sub 1</h4>
                    </div>
                    <div id="GroupCSub2" class="subgroup">
                        <h4>Group C Sub 2</h4>
                    </div>
                </section>
            </div>
        </div>

        <style type='text/css'>
/*        .group {
            background: yellow;
            width: 200px;
            height: 500px;
        }

        .group .subgroup {
            background: orange;
            width: 150px;
            height: 200px;
        }*/

        .fixed {
            position: fixed;
        }

        /* sidebar */
        .bs-docs-sidebar {
            padding-left: 20px;
            margin-top: 20px;
            margin-bottom: 20px;
        }

        /* all links */
        .bs-docs-sidebar .nav > li > a {
            color: #999;
            border-left: 2px solid transparent;
            padding: 4px 20px;
            font-size: 13px;
            font-weight: 400;
        }

        /* nested links */
        .bs-docs-sidebar .nav .nav > li > a {
            padding-top: 1px;
            padding-bottom: 1px;
            padding-left: 30px;
            font-size: 12px;
        }

        /* active & hover links */
        .bs-docs-sidebar .nav > .active > a,
        .bs-docs-sidebar .nav > li > a:hover,
        .bs-docs-sidebar .nav > li > a:focus {
            color: #563d7c;
            text-decoration: none;
            background-color: transparent;
            border-left-color: #563d7c;
        }

        /* all active links */
        .bs-docs-sidebar .nav > .active > a,
        .bs-docs-sidebar .nav > .active:hover > a,
        .bs-docs-sidebar .nav > .active:focus > a {
            font-weight: 700;
        }

        /* nested active links */
        .bs-docs-sidebar .nav .nav > .active > a,
        .bs-docs-sidebar .nav .nav > .active:hover > a,
        .bs-docs-sidebar .nav .nav > .active:focus > a {
            font-weight: 500;
        }

        /* hide inactive nested list */
        .bs-docs-sidebar .nav ul.nav {
            display: none;
        }

        /* show active nested list */
        .bs-docs-sidebar .nav > .active > ul.nav {
            display: block;
        }

        /*Header formatting */
        .MainContent {
            margin-top: 50px;
        }

        .group, .subgroup {
            padding-top: 50px;
            margin-top: -50px;
        }

        </style>
        @* <script>
        $('body').scrollspy({
        target: '.bs-docs-sidebar',
        offset: 40
        });
        </script>*@


    </body>
</html>
